File: scripts/add_cb_ur_tags.py
import argparse
import pysam
import pandas as pd


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--bam", required=True, help="输入 BAM")
    parser.add_argument("--tags", required=True, help="read_tags.tsv")
    parser.add_argument("--output", required=True, help="输出 BAM")
    parser.add_argument("--read-id-col", default="read_id")
    parser.add_argument("--cb-col", default="cell_id")
    parser.add_argument("--ur-col", default="umi_for_clustering")
    parser.add_argument("--cr-col", default="barcode_dual")
    parser.add_argument("--bc5-col", default="barcode_5p")
    parser.add_argument("--bc3-col", default="barcode_3p")
    parser.add_argument("--umi5-col", default="umi_primary")
    parser.add_argument("--umi3-col", default="umi_backup")
    parser.add_argument(
        "--keep-untagged",
        action="store_true",
        help="保留无法写入 CB/UR 的 alignments；默认按 Sockeye 逻辑丢弃。",
    )
    return parser.parse_args()


def load_tag_map(path, args):
    df = pd.read_csv(path, sep="\t", dtype=str).fillna("")
    required = [args.read_id_col, args.cb_col, args.ur_col]
    missing = [c for c in required if c not in df.columns]
    if missing:
        raise ValueError(f"缺少列: {missing}")

    df = df[df[args.read_id_col].str.strip() != ""].copy()
    df = df.drop_duplicates(subset=[args.read_id_col], keep="first")

    tag_map = {}
    for _, row in df.iterrows():
        rid = row[args.read_id_col].strip()
        cb = row[args.cb_col].strip()
        ur = row[args.ur_col].strip()
        if cb and ur:
            tag_map[rid] = {
                "CB": cb,
                "UR": ur,
                "CR": row.get(args.cr_col, "").strip(),
                "C5": row.get(args.bc5_col, "").strip(),
                "C3": row.get(args.bc3_col, "").strip(),
                "U5": row.get(args.umi5_col, "").strip(),
                "U3": row.get(args.umi3_col, "").strip(),
            }
    return tag_map


def main():
    args = parse_args()
    tag_map = load_tag_map(args.tags, args)

    bam = pysam.AlignmentFile(args.bam, "rb")
    out = pysam.AlignmentFile(args.output, "wb", template=bam)

    total = 0
    tagged = 0
    total_read_ids = set()
    tagged_read_ids = set()

    for aln in bam.fetch(until_eof=True):
        total += 1
        rid = aln.query_name
        total_read_ids.add(rid)
        if rid in tag_map:
            tags = tag_map[rid]
            aln.set_tag("CB", tags["CB"], value_type="Z")
            aln.set_tag("UR", tags["UR"], value_type="Z")
            for tag in ["CR", "C5", "C3", "U5", "U3"]:
                if tags.get(tag):
                    aln.set_tag(tag, tags[tag], value_type="Z")
            tagged += 1
            tagged_read_ids.add(rid)
            out.write(aln)
        elif args.keep_untagged:
            out.write(aln)

    bam.close()
    out.close()
    pysam.index(args.output)

    print(f"total_alignments\t{len(total_read_ids)}")
    print(f"tagged_alignments\t{len(tagged_read_ids)}")
    print(f"unmatched_alignments\t{len(total_read_ids - tagged_read_ids)}")
    print(f"total_alignment_records\t{total}")
    print(f"tagged_alignment_records\t{tagged}")
    print(f"unmatched_alignment_records\t{total - tagged}")


if __name__ == "__main__":
    main()

File: src/bin/add_cb_ur_tags.rs
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;

use anyhow::{Context, Result};
use clap::{ArgAction, Parser};
use rust_htslib::bam::{self, Read};

#[derive(Debug, Parser)]
#[command(
    version,
    about = "Add CB/UR and related Flora tags to BAM from read_tags.tsv"
)]
struct Cli {
    #[arg(long)]
    bam: PathBuf,

    #[arg(long)]
    tags: PathBuf,

    #[arg(long)]
    output: PathBuf,

    #[arg(long = "read-id-col", default_value = "read_id")]
    read_id_col: String,
    #[arg(long = "cb-col", default_value = "cell_id")]
    cb_col: String,
    #[arg(long = "ur-col", default_value = "umi_for_clustering")]
    ur_col: String,
    #[arg(long = "cr-col", default_value = "barcode_dual")]
    cr_col: String,
    #[arg(long = "bc5-col", default_value = "barcode_5p")]
    bc5_col: String,
    #[arg(long = "bc3-col", default_value = "barcode_3p")]
    bc3_col: String,
    #[arg(long = "umi5-col", default_value = "umi_primary")]
    umi5_col: String,
    #[arg(long = "umi3-col", default_value = "umi_backup")]
    umi3_col: String,
    #[arg(long = "keep-untagged", action = ArgAction::SetTrue)]
    keep_untagged: bool,
}

pub fn main() -> Result<()> {
    let cli = Cli::parse();
    let tag_map = load_tag_map(&cli)?;
    let mut bam =
        bam::Reader::from_path(&cli.bam).with_context(|| format!("open {}", cli.bam.display()))?;
    let header = bam::Header::from_template(bam.header());
    let mut out = bam::Writer::from_path(&cli.output, &header, bam::Format::Bam)
        .with_context(|| format!("create {}", cli.output.display()))?;

    let mut total = 0usize;
    let mut tagged = 0usize;
    let mut total_read_ids = HashSet::new();
    let mut tagged_read_ids = HashSet::new();
    for rec in bam.records() {
        let mut rec = rec?;
        total += 1;
        let rid = String::from_utf8_lossy(rec.qname()).to_string();
        total_read_ids.insert(rid.clone());
        if let Some(tags) = tag_map.get(&rid) {
            push_tag(&mut rec, b"CB", &tags.cb)?;
            push_tag(&mut rec, b"UR", &tags.ur)?;
            maybe_push_tag(&mut rec, b"CR", &tags.cr)?;
            maybe_push_tag(&mut rec, b"C5", &tags.c5)?;
            maybe_push_tag(&mut rec, b"C3", &tags.c3)?;
            maybe_push_tag(&mut rec, b"U5", &tags.u5)?;
            maybe_push_tag(&mut rec, b"U3", &tags.u3)?;
            tagged += 1;
            tagged_read_ids.insert(rid);
            out.write(&rec)?;
        } else if cli.keep_untagged {
            out.write(&rec)?;
        }
    }
    drop(out);
    drop(bam);
    bam::index::build(&cli.output, None, bam::index::Type::Bai, 1)?;
    println!("total_alignments\t{}", total_read_ids.len());
    println!("tagged_alignments\t{}", tagged_read_ids.len());
    println!(
        "unmatched_alignments\t{}",
        total_read_ids.len().saturating_sub(tagged_read_ids.len())
    );
    println!("total_alignment_records\t{total}");
    println!("tagged_alignment_records\t{tagged}");
    println!(
        "unmatched_alignment_records\t{}",
        total.saturating_sub(tagged)
    );
    Ok(())
}

#[derive(Clone)]
struct TagRow {
    cb: String,
    ur: String,
    cr: String,
    c5: String,
    c3: String,
    u5: String,
    u3: String,
}

fn load_tag_map(cli: &Cli) -> Result<HashMap<String, TagRow>> {
    let mut reader = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .from_path(&cli.tags)
        .with_context(|| format!("open {}", cli.tags.display()))?;
    let headers = reader.headers()?.clone();
    let idx = |name: &str| {
        headers
            .iter()
            .position(|h| h == name)
            .context(format!("missing column: {name}"))
    };
    let read_id_idx = idx(&cli.read_id_col)?;
    let cb_idx = idx(&cli.cb_col)?;
    let ur_idx = idx(&cli.ur_col)?;
    let cr_idx = idx(&cli.cr_col)?;
    let bc5_idx = idx(&cli.bc5_col)?;
    let bc3_idx = idx(&cli.bc3_col)?;
    let umi5_idx = idx(&cli.umi5_col)?;
    let umi3_idx = idx(&cli.umi3_col)?;

    let mut tag_map = HashMap::new();
    for row in reader.records() {
        let row = row?;
        let rid = clean(row.get(read_id_idx));
        if rid.is_empty() || tag_map.contains_key(&rid) {
            continue;
        }
        let cb = clean(row.get(cb_idx));
        let ur = clean(row.get(ur_idx));
        if cb.is_empty() || ur.is_empty() {
            continue;
        }
        tag_map.insert(
            rid,
            TagRow {
                cb,
                ur,
                cr: clean(row.get(cr_idx)),
                c5: clean(row.get(bc5_idx)),
                c3: clean(row.get(bc3_idx)),
                u5: clean(row.get(umi5_idx)),
                u3: clean(row.get(umi3_idx)),
            },
        );
    }
    Ok(tag_map)
}

fn clean(value: Option<&str>) -> String {
    value.unwrap_or("").trim().to_string()
}

fn push_tag(rec: &mut bam::Record, tag: &[u8; 2], value: &str) -> Result<()> {
    rec.update_aux(tag, bam::record::Aux::String(value))?;
    Ok(())
}

fn maybe_push_tag(rec: &mut bam::Record, tag: &[u8; 2], value: &str) -> Result<()> {
    if !value.is_empty() {
        push_tag(rec, tag, value)?;
    }
    Ok(())
}
